### ISSUE TITLE
v3.0 TODO

### ISSUE THREAD

**Issue by DesignByOnyx** [StackStorm/st2flow#149]
This is a living list of TODOs for the version 3 release. This list should be constantly updated as items are completed or new issues are discovered. We encourage anybody and everybody to leave comments about any glitches, bugs, etc. and respectfully ask that Kirill and myself be the only ones to edit this list directly. Thanks!

**Terminology**
- **HIGH PRIORITY** - as the name implies...
- **BIG EFFORT** - used to denote something that could take longer than a couple days

## Global

- [x] ~~HIGH PRIORITY - standardize test scripts and paradigms to work with st2web repos~~
- [x] ~~HIGH PRIORITY - update startup script to allow configurable port ([see this PR](https://github.com/StackStorm/st2web/pull/604))~~
- [ ] handle routing - load correct workflow (2 human-days)
    - This should be done last as using hard-coded dummy data allows us to work quickly right now.
- [ ] overall aesthetic and UX review (Adri) (5 days)
- [x] current user indicator (with log out, etc) (1 day)
- [ ] make use of workflow validation api (.5 day, postpone)
- [ ] BIG EFFORT - workflow execution view (postpone)
- [ ] HIGH PRIORITY - make tests pass in CI (.5 day)
- [ ] write UI tests (5 days)
- [ ] cover UI part in type definitions (1 day)
- [ ] compile an updated icon set (Adri)

## Model

- [ ] HIGH PRIORITY - Any changes to transitions should supply both the changed transition AND the entire set of transitions for the parent ("from") task. (1 day)
    - It is impossible to cross reference a single transition from the common `TransitionInterface` to a model-specific transition (mainly with orquesta). See [this comment below](#issuecomment-426107911) more in depth explanation.
- [x] ~~Use schema validation to ensure yaml is what we expect ([done in this PR](https://github.com/StackStorm/st2flow/pull/152))~~

- [x] HIGH PRIORITY - Allow *new* comments to be written back to yaml
- [x] Infer formatting and stylistic choices from existing tokens when adding new tasks, transitions, etc.
  - [ ] validate we're happy with the way style is followed

- [x] Implement metadata model
  - [x] HIGH PRIORITY - fix JSON mode in st2flow-yaml (workaround)
  - [ ] Sync with core team to implement file access api

- [ ] Implement transition update\delete API for orchestra
- [x] Add ability to rename tasks
- [ ] Algorithm to determine "success" or "fail" with orquesta transitions
- [ ] Transition data comments (transition color, position of a start and end point on the task)

- [ ] BIG EFFORT - Implement Mistral model (2 days)
- [ ] BIG EFFORT - Mistral to Orchestra conversion (postpone, 4 days)

## Canvas

- [x] ~~visual indications/styles for selected task~~
- [ ] transition selection (1 day)
- [ ] action button behaviors (undo, redo, et al) (.5 day)
- [ ] smart placement of tasks which have no coordinates
  - [ ] port existing algorithm (1 day)
  - [ ] determine if our existing algorithm still holds
  - [ ] take into account multiple connection points

- [ ] add pack icons to tasks (2 hours)
- [ ] allow replacing icons with emoticons (pack icons provide little information) (postpone)
- [ ] **BIG EFFORT** allow resizing tasks (postpone)
  - [ ] make sure smart placement and collision detection algorithms are aware of that change

- [ ] transition appearance tweaking
  - [ ] gracefully ignore hardcoded values when tasks are too close together (1 day)
  - [ ] improve algorithm of calculation "step" position depending on tasks relative position
  - [ ] synchronize task "step" position of multiple transitions for better appearance
- [ ] drag and drop - transition arrow should follow box while being dragged _(could have severe performance penalties)_ (.5 day naive implementation + whatever it takes to make it quick) (postpone)
- [ ] in the same vein, transition arrow should follow cursor when you drag the handle or move it over another's task handle (^^)
- [ ] detect\prevent crossing of transition lines (postpone, review during UX check)
- [ ] bypass other tasks and leave enough clearance for better visual appeal (^^)
- [ ] custom styling (color, dashing, dotting, outlines, midpoint icons)

This two are most likely related to model not updating properly and need reevaluation once we done with coordinates in comments:
- [x] drag and drop - item shifts on second drag
    - upon second click, item jumps to original position before moving

## Editor

- [x] ~~HIGH PRIORITY - handle YAML errors gracefully~~
    - [x] ~~prevent hard crash~~
    - [x] ~~meaningful, concise error messages (need UX input)~~
    ~~([fixed in this PR](https://github.com/StackStorm/st2flow/pull/152))~~
- [x] ~~HIGH PRIORITY - allow yaml scrolling ([scrolling fixed here](https://github.com/StackStorm/st2flow/pull/148/commits/7e82954f805515aaaf0295cd78f1e6066553d49c))~~
- [ ] Show yaml errors in editor (.5 day)
- [ ] Add search (postpone)
- [ ] Show selected task (1 day)

- [ ] **BIG EFFORT** Update editor to new version (postpone until further notice)

### Details Form

- [x] update form based on selected task
- [x] handle form change events (needs meta model)
- [x] add text editor for metadata (.5 day)
- [ ] finish work on action parameters form (1 day)
    
## Palette

- [ ] HIGH PRIORITY - dropping a task/workflow onto the canvas should create a new task
- [x] ~~HIGH PRIORITY - allow scrolling ([scrolling fixed here](https://github.com/StackStorm/st2flow/pull/148/commits/7e82954f805515aaaf0295cd78f1e6066553d49c))~~
